### ISSUE TITLE
Improve semantic annotation of document identifiers

### ISSUE THREAD

**Issue by opoudjis** [metanorma/metanorma-iso#1236]
Document identifiers are semantically tagged by their components under spans—publisher, year, number, part number: this is a requirement for DOC styling. The refactoring of document identifiers for https://github.com/metanorma/isodoc/issues/606 has moved this functionality to Presentation XML, which is OK—but it also exposes errors in inferred markup. This is a best effort guess, but I do want to improve the guess:

````
               <span class="stdpublisher">ISO </span>
               <span class="stddocNumber">16634</span>
               :--
               <span class="stdpublisher">(all </span>
               <span class="stdpublisher">parts)</span>


               <span class="stdpublisher">ISO </span>
               <span class="stddocNumber">20483</span>
               :
               <span class="stdyear">2013</span>
               -
               <span class="stddocPartNumber">2014</span>
````

**Comment by ronaldtse** [comment]
Since this new presentation xml will differ greatly, please help make sure we use tight version binds "~>" for minimum versions of isodoc and mn2pdf gems.

**Comment by ronaldtse** [comment]
"(all parts)" is supported in both Relaton and Pubid gems.

**Comment by opoudjis** [comment]
Since this is not a priority concern—it only impacts Word output in ISO and ITU, I am going to (a) move the current code annotating identifiers out of Presentation XML after the xrefs-refactor, and to the DOC branch, and (b) get pub-id to implement it.

**Comment by opoudjis** [comment]
> Since this new presentation xml will differ greatly, please help make sure we use tight version binds "~>" for minimum versions of isodoc and mn2pdf gems.

isodoc is found to standoc which is bound to everything else. These will be minor version updates to isodoc.

mn2pdf is also bound to isodoc, and until now has had only a major version dependency. @Intelligent2013 I assume that we are going to need to bind to more granular versions of mn2pdf for the duration of this exercise.

**Comment by opoudjis** [comment]
This issue goes on hold pending https://github.com/metanorma/pubid/issues/14

**Comment by opoudjis** [comment]
Unblocked

**Comment by opoudjis** [comment]
Need to preempt any nbsp inserted into identifiers, they are preventing the proper parsing of the identifier under pubid.

**Comment by opoudjis** [comment]
Also need to port IEEE, IEC, JIS, BSI

**Comment by opoudjis** [comment]
## Behaviour adjustment: round-trip integrity check on `Pubid::Registry.parse`

Surfaced while extending the semantic-docidentifier-annotation refactor (this issue) to IEEE.

### The bug

`Pubid::Registry.parse` is over-eager: any input that vaguely matches "letters
followed by digits" is parsed as an IEEE Std identifier, with the parser
**inventing** an `IEEE Std ` prefix and labelling the input itself as a
`docnumber`. Verified at `pubid` 1.15.17 / `pubid-ieee` 1.15.17:

```ruby
Pubid::Registry.parse("REF4").to_s
# => "IEEE Std REF4"     # round-trip is NOT identity
Pubid::Registry.parse("REF4").to_s(annotated: true)
# => "<span class=\"publisher\">IEEE</span> Std <span class=\"docnumber\">REF4</span>"

Pubid::Registry.parse("ABC123").to_s
# => "IEEE Std ABC123"   # same hallucination
```

This bit us in `metanorma-ieee`'s spec suite: bibitem identifiers like
`REF1`, `REF4` were being rendered as `IEEE Std REF4` with `IEEE` wrapped
as `<span class="std_publisher">` and `REF4` wrapped as `<span
class="std_docNumber">`. The "IEEE Std" string was synthesised from thin
air — not present anywhere in the input docidentifier.

Bracketed refs (`[REF4]`) are unaffected — Pubid rejects them outright with
a `ParseError`, which routes them through the regex fallback (or, with the
fix below, returns the input unchanged).

### Generic Pubid risk

This is not specific to IEEE or to this refactor. Any consumer of
`Pubid::Registry.parse(...).to_s(annotated: true)` will silently inject
phantom publisher/std tokens for any docid-like-but-not-actually-a-docid
input. ISO's existing usage of `Pubid::Registry` on
`fix/improved-pubid-annotation` has the same exposure; it just doesn't
surface in ISO specs because ISO bibitem refs are typically bracketed
(`[REF4]`) and so hit the `ParseError` path.

### Fix

Single-line round-trip integrity check in the shared
`std_docid_semantic_parse` orchestrator at
`isodoc/lib/isodoc/presentation_function/docid_semantic.rb`:

```ruby
def std_docid_semantic_parse(id)
  parsed = Pubid::Registry.parse(id)
  return id if parsed.to_s != id   # Pubid invented structure not in input
  parsed.to_s(annotated: true)
rescue Pubid::Core::Errors::ParseError
  std_docid_semantic_full(id)
end
```

If `parsed.to_s` (without `annotated:`) does not equal the original input
verbatim, Pubid has reconstructed (i.e. hallucinated) structure and we
return the input unchanged rather than emit a misleading annotation.

This lives in the base, so it benefits every flavor that opts in
(`annotate_docid? = true`) — ISO, IEEE, and any future opt-in flavor. No
flavor-specific code is needed.

### Spec impact

After applying this guard, many of the IEEE spec diffs that initially
looked like "publisher inside vs outside the span" rebaselines were in
fact this hallucination bug rather than legitimate output changes. Specs
that referenced bibitem IDs like `REF1`/`REF4` revert to the original
expected output. I will run the IEEE suite under the guarded base next and
report the remaining (genuine) rebaseline scope on the IEEE PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by opoudjis** [comment]
## Incidental bugfix: regex fallback no longer drops trailing characters

Surfaced while rebaselining IEEE specs under the hoisted base.

The original IEEE `std_docid_semantic_full` regex captured only `text +
docnumber [+ year]` and silently discarded everything after the docnumber
in the input. This dropped legitimate trailing tokens such as the
incomplete-date marker `:--` (e.g. `ISO 16634:--` rendered as
`ISO 16634`, losing the `:--`).

The hoisted base regex uses `.sub` (one-shot per token) over the input
string, which leaves any text past the matched tokens intact. So
`ISO 16634:--` now renders as
`<span class="publisher">ISO</span> <span class="docnumber">16634</span>:--`
— the `:--` is preserved as plain trailing text.

Fixtures expecting the old lossy output need to be rebaselined. No
behaviour change required for code outside the test fixtures.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
